specs/ChainSpec/Consensus.md

```diff
@@ -120,22 +118,35 @@ def get_approvals(self, target_height):
 
 A block producer assigned for a particular height produces a block at that height whenever they have `get_approvals` return approvals from block producers whose stake collectively exceeds 2/3 of the total stake.
 
-## Safety
+## Epoch Switches
+There's a parameter `epoch_length` in genesis config that defines the expected length of an epoch. Say a particular epoch `e_cur` started at height `h`, and say the next epoch will be `e_next`. Say `BP(e)` is a set of block producers in epoch `e`. Say `LFH(B)` is the height of most recent final block in the ancestry of `B`. The following are the rules of what blocks contain approvals from what block producers, and belong to what epoch.
 
-We call approvals for which `prev_height + 1 = target_height` endorsements, and all other approvals skips.
+- Any block `B` with height between `h` and `h + epoch_length - 3` (both inclusive) is in the epoch `e_cur` and must have approvals from more than 2/3 of `BP(e_cur)` (stake-weighted).
+- Any block `B` with height `h + epoch_length - 2` or higher for which `LFH(prev(B)) < h + epoch_length - 3` is in the epoch `e_cur` and must have the approvals from both more than 2/3 of `BP(e_cur)` and more than 2/3 of `BP(e_next)` (both stake-weighted).
+- The first block `B` with height `h + epoch_length - 2` or higher and `LFH(prev(B)) >= h + epoch_length - 3` is in the epoch `e_next` and only needs the approvals from more than 2/3 of `BP(e_next)` (stake-weighted).
```

Comment:
This means that an epoch only has `epoch_length - 2` blocks under normal circumstances, which seems suboptimal.

Reply:
Incorrect, this makes most epochs have exactly `epoch_length` blocks.

Let's say some block at height `h'` triggers the last condition. The the block at height `h'-1` has a final block with height `h + epoch_length - 3`. Since the final block is 2 heights before the block that observes it, it means that `h'-1-2 = h + epoch_length - 3`, or `h' = h + epoch_length`.

Reply:
👍